### ISSUE TITLE
`new Polly` errors not caught

### ISSUE THREAD

**Issue by swashcap** [gribnoysup/setup-polly-jest#5]
### Problem

I am observing that `new Polly(/* ... */)` errors in setup-polly-jest aren't caught here:

https://github.com/gribnoysup/setup-polly-jest/blob/8c0acedca7f84a788f80974460893605f335a946/src/setupJasmine.js#L101

Instead, an error occurs when `pollyContext.polly.stop()` is called: `pollyContext.polly` is null:

```
  console.error jest/setup.js:95
    TypeError: Cannot read property 'stop' of null
        at Object.<anonymous> (/Users/swashcap/project/node_modules/setup-polly-jest/lib/setupJasmine.js:153:38)
        at Generator.next (<anonymous>)
        at asyncGeneratorStep (/Users/swashcap/project/node_modules/setup-polly-jest/lib/setupJasmine.js:8:103)
        at _next (/Users/swashcap/project/node_modules/setup-polly-jest/lib/setupJasmine.js:10:194)
        at /Users/swashcap/project/node_modules/setup-polly-jest/lib/setupJasmine.js:10:364
        at new Promise (<anonymous>)
        at Object.<anonymous> (/Users/swashcap/project/node_modules/setup-polly-jest/lib/setupJasmine.js:10:97)
        at Object.after (/Users/swashcap/project/node_modules/setup-polly-jest/lib/setupJasmine.js:161:25)
        at resolve (/Users/swashcap/project/node_modules/jest-jasmine2/build/queueRunner.js:43:12)
        at new Promise (<anonymous>)
        at mapper (/Users/swashcap/project/node_modules/jest-jasmine2/build/queueRunner.js:26:19)
        at promise.then (/Users/swashcap/project/node_modules/jest-jasmine2/build/queueRunner.js:73:41)
        at process._tickCallback (internal/process/next_tick.js:68:7)

```

### Versions

* setup-polly-jest: `setup-polly-jest@0.5.2`
* Polly: `@pollyjs/core@2.6.2`
* npm: `6.9.0`
* Node.js: `10.16.0`

**Comment by gribnoysup** [comment]
Hi @swashcap! Thanks for reporting this. Can you help me to understand why it's failing for you better by creating a reproduction example?

**Comment by gribnoysup** [comment]
Hey @swashcap I think I could reproduce it after a bit of digging into the issue. I published the fix  as `setup-polly-jest@0.5.3-issue-5.0`, it should fail tests with proper errors now.

```sh
npm install setup-polly-jest@0.5.3-issue-5.0
```

Can you please try it and tell me if it's working for you?

**Comment by swashcap** [comment]
@gribnoysup Thanks for looking into this! I'll make a little test case and confirm it works.

**Comment by swashcap** [comment]
Just created https://github.com/swashcap/setup-polly-jest-constructor-error. I'm seeing this error on the `setup-polly-jest@0.5.3-issue-5.0` test:

```
 FAIL  setup-polly-jest-0.5.3-issue-5.0/index.test.js
  ● Test suite failed to run

    Cannot find module 'core-js/modules/es.symbol.description' from 'setupJasmine.js'

    However, Jest was able to find:
    	'lib/setupJasmine.js'

    You might want to include a file extension in your import, or update your 'moduleFileExtensions', which is currently ['js', 'json', 'jsx', 'ts', 'tsx', 'node'].

    See https://jestjs.io/docs/en/configuration#modulefileextensions-array-string

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:259:17)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.546s
Ran all test suites matching /setup-polly-jest-0.5.3-issue-5.0\/index.test.js/i.
error Command failed with exit code 1.
```

The 0.5.2 test runs without this error. 🤔

**Comment by gribnoysup** [comment]
Sorry for that, I messed up the build 🤦‍♂  Published a fixed one as `setup-polly-jest@0.5.3-issue-5.2`

**Comment by swashcap** [comment]
Just bumped in [swashcap/setup-polly-jest-constructor-error](https://github.com/swashcap/setup-polly-jest-constructor-error). Unfortunately, `0.5.3-issue-5.2` emits the same `TypeError: Cannot read property 'configure' of null` error as `0.5.2`.

**Comment by gribnoysup** [comment]
Hmm, just cloned the repo and tried it myself. Not sure if we are talking about the same errors or not... 🤔 So when I run tests with `0.5.2`, this is the output that I'm getting:

```
> jest --runInBand "setup-polly-jest-0.5.2/index.test.js"

(node:9459) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'stop' of null
(node:9459) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (reje
ction id: 3)
(node:9459) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
 FAIL  setup-polly-jest-0.5.2/index.test.js
  sample
    ✕ assertion (120ms)

  ● sample › assertion

    PollyError: [Polly] Persister matching the name `fs-whoops` was not registered.

      at Object.assert (node_modules/@pollyjs/utils/src/utils/assert.js:5:11)

  ● sample › assertion

    TypeError: Cannot read property 'configure' of null

      27 | 
      28 |   test("assertion", async () => {
    > 29 |     context.polly.configure({
         |                   ^
      30 |       recordIfMissing: true
      31 |     });
      32 | 

      at Object.configure (setup-polly-jest-0.5.2/index.test.js:29:19)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        0.797s, estimated 1s
```

There are indeed some async errors that were not caught:

```
(node:9459) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'stop' of null
(node:9459) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (reje
ction id: 3)
(node:9459) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

But when I'm running with the issue-5 branch this is the output:

```
> jest --runInBand "setup-polly-jest-0.5.3-issue-5.2/index.test.js"

 FAIL  setup-polly-jest-0.5.3-issue-5.2/index.test.js
  sample
    ✕ assertion (118ms)

  ● sample › assertion

    PollyError: [Polly] Persister matching the name `fs-whoops` was not registered.

      at Object.assert (node_modules/@pollyjs/utils/src/utils/assert.js:5:11)

  ● sample › assertion

    TypeError: Cannot read property 'configure' of null

      27 | 
      28 |   test("assertion", async () => {
    > 29 |     context.polly.configure({
         |                   ^
      30 |       recordIfMissing: true
      31 |     });
      32 | 

      at Object.configure (setup-polly-jest-0.5.3-issue-5.2/index.test.js:29:19)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        0.818s, estimated 1s
```

No uncaught exceptions anymore.

Are you getting the same output? Is there something else that I'm missing?

**Comment by swashcap** [comment]
> Are you getting the same output? Is there something else that I'm missing?

Re-looking, and I get the same test output. I wasn't reading it correctly.

I think this causes some confusion:

```
    TypeError: Cannot read property 'configure' of null

      27 | 
      28 |   test("assertion", async () => {
    > 29 |     context.polly.configure({
         |                   ^
      30 |       recordIfMissing: true
      31 |     });
      32 | 

      at Object.configure (setup-polly-jest-0.5.3-issue-5.2/index.test.js:29:19)
```

It's the last test that fails, and the reason it fails is a few lines above, which I must have ignored (whoops!). Would it be possible to bail on the failed `describe` block when `new Polly(/* ... */)` throws, or is that more of a Jest user setting?

**Comment by gribnoysup** [comment]
Ahh, okay, that makes sense now. Yeah, you're right, it can be confusing.

So the technical reason for this is that Polly instance is not created when `seupPolly` is called. This happens later, when a "before spec" callback for a spec is triggered by jest. Jest doesn't stop execution if the before hook fails, it just goes forward with the tests.

My guess is that if you would try running jest with `--bail` option, this would not be the case, but I don't know for sure, I haven't tried it.

I'm not sure I can change this and prevent jest from continuing with the spec, but I guess this library can at least throw a more meaningful error when trying to access Polly when it doesn't exist 🤔

**Comment by swashcap** [comment]
It looks like you might potentially have access to the `new Polly` error in the closure. Maybe something like this?

```js
spec.beforeAndAfterFns = () => {
  let maybeError;

  // ...

  before = () => {
    // ...

    try {
      new Polly({ /* ... */ })
    } catch (error) {
      maybeError = error
      throw error
    }
  }

  after = (done) => {
    // ...

    // Fail here with some context:
    if (!pollyContext.polly && maybeError) {
      return done.fail(`setup-polly-jest's `setupPolly()` call failed with error:

${maybeError.stack}

Please check your Polly.JS configuration.`)
    }

    try {
      await pollyContext.polly.stop()
    } catch (error) {
      done.fail(error)
     }
  }
}
```

**Comment by gribnoysup** [comment]
Ooh, nice! I love the idea of storing the error from `before` hook 👍

**Comment by gribnoysup** [comment]
Just published `setup-polly-jest@0.5.3-issue-5.4`. Do you mind giving it another go and telling me what you think before I release it?

**Comment by swashcap** [comment]
Just upgraded in swashcap/setup-polly-jest-constructor-error@ab715ee3503f24a4796b4bafdb36f11a2482b617:

> ```shell
> PollyError: [Polly] Persister matching the name `fs-whoops` was not registered. Check `setupPolly` method and make sure it's configured correctly.
>
>   at Object.assert (node_modules/@pollyjs/utils/src/utils/assert.js:5:11)
> ```

🎉 

Great work!

**Comment by gribnoysup** [comment]
Released as `setup-polly-jest@0.6.0`. Thanks for helping with testing the changes!
